tests/test_routes.py

```diff
@@ -1,8 +1,25 @@
+import pathlib
 import unittest
-from unittest.mock import patch
 
-from webapp.app import app
-from webapp.models import CVE, Notice, Status
+from alchemy_mock.mocking import UnifiedAlchemyMagicMock
+from flask import json
+
+from tests.auth_mock import mock_auth_decorator
+from tests.alchemy_mock_data import data
+from tests.database_mock import (
+    mocked_inspector,
+    mocked_release_codenames,
+    mocked_status_statuses,
+)
+from webapp import auth, database
+
+auth.authorization_required = mock_auth_decorator()
+database.db_session = UnifiedAlchemyMagicMock(data=data)
+database.inspector = mocked_inspector
+database.release_codenames = mocked_release_codenames
+database.status_statuses = mocked_status_statuses
```

Comment:
This would normally be done in a [setUp method](https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUp). Is there a reason why you've made the session global?

Reply:
Yes. Because `app` gets instantiated when I'm importing the global `app`.

I need `app` to use the mocked database values when it gets instantiated, therefore I am forced to put them above.

One way I can see this could be moved inside `setUp` is if we take the modular approach of bootstrapping the app, with the `create_app` function.

Or I can import the `app` inside the `setUp`, but it's still not ideal..

Reply:
I can see the issue here, but you are monkey patching the database object and the auth decorator before importing the app object, I think that makes clear that we might need a bit of refactoring.

I think here we need to use `app.config['TESTING']` and when is `True` change the decorator behaviour and provide the mocked db session. I would like to see what @nottrobin thinks

Reply:
I don't like the sound of changing the behaviour of a decorator based on `testing = true`.

I'm sure I'm missing a bunch of context here. Yes you're using a global `app` objects, but [the `app` itself](https://github.com/canonical-web-and-design/ubuntu-com-security-api/blob/master/webapp/app.py#L17) doesn't seem to know anything about the DB does it? Instead the global db_session is used directly [in views](https://github.com/canonical-web-and-design/ubuntu-com-security-api/blob/master/webapp/views.py#L28).

So if you want a mock database, it seems like you should be creating your own db_session (rather than using [the global one](https://github.com/canonical-web-and-design/ubuntu-com-security-api/blob/master/webapp/database.py#L9)) shouldn't you? Can't you instantiate this special mock DB session in `setUp`?

Reply:
Oh we're talking about different things. I'm talking about DB stuff, you're talking about `authorization_required`

Reply:
I think the right way to test auth would be to create a mock macaroonbakery. This way we can test all of our code properly, and simply swap out the external macaroon library for a mock one.

Might be worth asking Martin Hilton or someone about if they have any mock macaroon bakery tooling. Or we could just create the mock ourselves.

Reply:
For now it's fine to override `authorization_required` this way, but I think in the long term we should find a way to pass through mock versions of all external libraries to the real `authorization_required` function, so that we're actually testing our auth code.

Reply:
There are two "monkey patches" happening before importing the global app: authentication and database.
They both need to happen outside the `setUp` (before importing `app`), and I will go over each situation separately:
1) authentication: I think it has to do with how flask registers decorated views when you instantiate the app. Overriding the auth inside `setUp` is already too late. The app will use the global auth for tests.
2) database: Please correct me if my python knowledge is wrong, but this is what I found. In `app.py` we register views in the routes. Which makes us import `views.py` in `app.py`. In `views.py` we import from `database.py` `db_session`, as we make use of it inside the views. The app learns and sets all the `database.py` values the moment it gets imported, so even before `app` gets instantiated in `app.py`. Similarly to `auth`, it's too late if I override any of those `database.py` values inside `setUp`.

Reply:
> So if you want a mock database, it seems like you should be creating your own db_session (rather than using [the global one](https://github.com/canonical-web-and-design/ubuntu-com-security-api/blob/master/webapp/database.py#L9)) shouldn't you? Can't you instantiate this special mock DB session in `setUp`?

It's possible to instantiate it on the `setUp` but then how you make the views use this session? Right now it's not possible unless you do what Albert is doing here, right?

Reply:
Damn you're right.

This is the trouble with globals. I do think in the long run we should think about how to restructure this stuff so we can mock it well (dependency injection). I was already kinda worried about the global DB session.

But go ahead and leave this as it is for now. Sorry for delaying things.